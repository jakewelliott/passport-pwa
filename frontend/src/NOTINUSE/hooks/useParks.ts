import { useQuery } from '@tanstack/react-query';
import type { Park } from './types';
import { api, type APIError } from './api';

export const parkKeys = {
  all: ['parks'] as const,
  lists: () => [...parkKeys.all, 'list'] as const,
  detail: (id: number) => [...parkKeys.all, 'detail', id] as const,
};

export const useParks = () => {
  return useQuery<Park[], APIError>({
    queryKey: parkKeys.lists(),
    queryFn: () => api.getParks(),
  });
};

export const usePark = (id: number) => {
  return useQuery<Park, APIError>({
    queryKey: parkKeys.detail(id),
    queryFn: () => api.getPark(id),
    enabled: !!id,
  });
};
