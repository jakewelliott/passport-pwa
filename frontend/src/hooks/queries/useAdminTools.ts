import {
    API_AUTH_URL,
    API_BUCKET_LIST_URL,
    API_PARKS_URL,
    API_TRAILS_URL,
    fetchDelete,
    fetchPost,
    fetchPut,
} from '@/lib/fetch';
import type { BucketListItem, Park, Trail, UserProfile } from '@/types';
import { useMutation } from '@tanstack/react-query';
import type { Cruddable } from '../useCRUD';

// ADAM: normally i think one file, one purpose, but this is a collection of admin tools
// so i think it's less ugly to have all in one file

// parks
export const useCreatePark = () => {
    return useMutation({
        mutationKey: ['admin', 'parks', 'create'],
        mutationFn: (park: Park) => fetchPost(API_PARKS_URL, park),
    });
};

export const useUpdatePark = () => {
    return useMutation({
        mutationKey: ['admin', 'parks', 'update'],
        mutationFn: (park: Park) => fetchPut(`${API_PARKS_URL}/${park.id}`, park),
    });
};

export const useDeletePark = () => {
    return useMutation({
        mutationKey: ['admin', 'parks', 'delete'],
        mutationFn: (park: Park) => fetchDelete(`${API_PARKS_URL}/${park.id}`),
    });
};

export const useParkCRUD: Cruddable<Park> = {
    createHook: useCreatePark,
    updateHook: useUpdatePark,
    deleteHook: useDeletePark,
};

// bucket list
export const useCreateBucketListItem = () => {
    return useMutation({
        mutationKey: ['admin', 'bucketList', 'create'],
        mutationFn: (bucketListItem: BucketListItem) => fetchPost(API_BUCKET_LIST_URL, bucketListItem),
    });
};
export const useUpdateBucketListItem = () => {
    return useMutation({
        mutationKey: ['admin', 'bucketList', 'update'],
        mutationFn: (bucketListItem: BucketListItem) =>
            fetchPut(`${API_BUCKET_LIST_URL}/${bucketListItem.id}`, bucketListItem),
    });
};

export const useDeleteBucketListItem = () => {
    return useMutation({
        mutationKey: ['admin', 'bucketList', 'delete'],
        mutationFn: (bucketListItem: BucketListItem) => fetchDelete(`${API_BUCKET_LIST_URL}/${bucketListItem.id}`),
    });
};

export const useBucketListCRUD: Cruddable<BucketListItem> = {
    createHook: useCreateBucketListItem,
    updateHook: useUpdateBucketListItem,
    deleteHook: useDeleteBucketListItem,
};

// trails
export const useCreateTrail = () => {
    return useMutation({
        mutationKey: ['admin', 'trails', 'create'],
        mutationFn: (trail: Trail) => fetchPost(API_TRAILS_URL, trail),
    });
};

export const useUpdateTrail = () => {
    return useMutation({
        mutationKey: ['admin', 'trails', 'update'],
        mutationFn: (trail: Trail) => fetchPut(`${API_TRAILS_URL}/${trail.id}`, trail),
    });
};

export const useDeleteTrail = () => {
    return useMutation({
        mutationKey: ['admin', 'trails', 'delete'],
        mutationFn: (trail: Trail) => fetchDelete(`${API_TRAILS_URL}/${trail.id}`),
    });
};

export const useTrailCRUD: Cruddable<Trail> = {
    createHook: useCreateTrail,
    updateHook: useUpdateTrail,
    deleteHook: useDeleteTrail,
};

// users

export const useDontCallMe = () => {
    // ADAM: this is a no-op for type safety
    // i normally would never do this but here we are
    return useMutation({
        mutationKey: ['admin', 'users', 'update'],
        mutationFn: (user: UserProfile) => Promise.reject(new Error('NEVER EVER EVER CALL THIS FUNCTION')),
    });
};

export const useUpdateUserRole = () => {
    return useMutation({
        mutationKey: ['admin', 'users', 'updateRole'],
        mutationFn: (user: UserProfile) => fetchPut(`${API_AUTH_URL}/${user.id}/role`, user),
    });
};

// ADAM: yeah and ill figure this one out later
export const useUpdateUserPassword = () => {
    return useMutation({
        mutationKey: ['admin', 'users', 'updatePassword'],
        mutationFn: (user: UserProfile) => fetchPut(`${API_AUTH_URL}/${user.id}/password`, user),
    });
};

export const useUserCRUD: Cruddable<UserProfile> = {
    createHook: useDontCallMe,
    updateHook: useUpdateUserRole,
    deleteHook: useDontCallMe,
};
