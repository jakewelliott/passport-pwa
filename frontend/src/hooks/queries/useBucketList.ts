import { dbg } from '@/lib/debug';
import { API_BUCKET_LIST_URL, API_COMPLETED_BUCKET_LIST_ITEMS_URL, fetchGet, fetchPost } from '@/lib/fetch';
import type { BucketListCompletion, BucketListItem } from '@/types';
import { useMutation, useQuery } from '@tanstack/react-query';
import { useLocation } from '../useLocation';

/** Gets all bucket list items for all parks */
const useBucketListItems = () => {
  return useQuery({
    queryKey: ['bucketListItems'],
    queryFn: () => fetchGet(API_BUCKET_LIST_URL),
  });
};

/** Gets all completed bucket list items for a user */
const useCompletedBucketListItems = () => {
  return useQuery({
    queryKey: ['completedBucketListItems'],
    queryFn: () => fetchGet(API_COMPLETED_BUCKET_LIST_ITEMS_URL),
  });
};

/** Toggles the completion status of a bucket list item for a user
 *  Call mutate with the itemId to toggle the completion status
 */
const useToggleCompletion = () => {
  const { refetch } = useCompletedBucketListItems();
  const { geopoint } = useLocation();

  dbg('MISC', 'utc', geopoint);

  return useMutation({
    mutationFn: async (itemId: number) => {
      const response = await fetchPost(`${API_BUCKET_LIST_URL}/${itemId}`, {
        latitude: geopoint?.latitude,
        longitude: geopoint?.longitude,
        inaccuracyRadius: geopoint?.accuracy,
      });
      refetch();
      return response;
    },
  });
};

/** Call with parkId as undefined to get all bucket list items and completed items */
export const useBucketList = (parkId?: number) => {
  const { data: allItems, isLoading: isLoadingAllItems } = useBucketListItems();
  const { data: allCompletedItems, isLoading: isLoadingCompletedItems } = useCompletedBucketListItems();
  const { mutate: toggleCompletion, isPending: isMarkingAsComplete } = useToggleCompletion();

  const isLoading = isLoadingAllItems || isLoadingCompletedItems || isMarkingAsComplete;

  // filtering logic
  const parkItems = allItems?.filter((item: BucketListItem) => item.parkId === parkId);
  const parkCompletedItems = allCompletedItems?.filter((item: BucketListCompletion) => item.itemId === parkId);

  return {
    items: parkId === undefined ? allItems : parkItems,
    completed: parkId === undefined ? allCompletedItems : parkCompletedItems,
    toggleCompletion,
    isLoading,
  };
};
