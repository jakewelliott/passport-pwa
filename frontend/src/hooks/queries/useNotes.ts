import { dbg } from '@/lib/debug';
import { API_NOTES_URL, fetchGet, fetchPost } from '@/lib/fetch';
import type { ParkNote } from '@/types';
import { useMutation, useQuery } from '@tanstack/react-query';

export const useNote = (parkId: number) => {
  dbg('HOOK', 'useNote', parkId);

  return useQuery<ParkNote>({
    queryKey: ['notes', parkId],
    queryFn: () => fetchGet(`${API_NOTES_URL}/${parkId}`),
  });
};

export const useGetAllNotes = () => {
  dbg('HOOK', 'useGetAllNotes');

  return useQuery<ParkNote[]>({
    queryKey: ['notes'],
    queryFn: async () => await fetchGet(API_NOTES_URL),
  });
};

export const useUpdateNote = () => {
  const { refetch } = useGetAllNotes();
  return useMutation({
    mutationFn: ({ parkId, note }: { parkId: number; note: string }) =>
      fetchPost(`${API_NOTES_URL}/${parkId}`, { note: note, updatedAt: new Date() }),
    onSuccess: () => {
      dbg('HOOK', 'useUpdateNote', 'refetching all notes...');
      refetch();
    },
  });
};
