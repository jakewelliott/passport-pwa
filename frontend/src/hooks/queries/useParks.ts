import { useQuery } from '@tanstack/react-query';
import type { Park, ParkAbbreviation } from '@/lib/mock/types';
import { api } from '@/lib/mock/api';

// ADAM:
// This is a very simple query hook.
// It's a good starting point for most queries.
// It's a good idea to have a query hook for each endpoint.

// useQuery returns an object with some useful properties:
// - data: the data returned from the query
// - isLoading: a boolean that indicates if the query is loading (data is undefined)
// - isError: a boolean that indicates if the query has an error
// - error: the error returned from the query

export const useParks = () => {
  return useQuery<Park[]>({
    queryKey: ['parks'],
    queryFn: (): Park[] => api.getParks(),
  });
};

export const usePark = (code: ParkAbbreviation) => {
  return useQuery<Park>({
    queryKey: ['park', code],
    queryFn: (): Park => api.getPark(code),
    enabled: !!code,
  });
};
