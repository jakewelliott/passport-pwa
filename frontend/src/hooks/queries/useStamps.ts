import type { ParkAbbreviation } from '@/lib/mock/types';
import { useUser } from './useUser';
import { api } from '@/lib/mock/api';
import { useQuery } from '@tanstack/react-query';
/**
 * Gets all of the local user's stamps, empty array if user is not loaded
 * @returns The stamps for the user
 */
export const useStamps = () => {
  // re-use our query hooks whenever possible
  const { data: user } = useUser();
  return useQuery({
    queryKey: ['stamps', user?.id],
    queryFn: () => api.getUserStampsByID(user?.id || 0),
  });
};

/**
 * Gets a single stamp for a user, null if stamp not collected yet
 * @param code The code of the park to get the stamp for
 * @returns The stamp for the user
 */
export const useStamp = (code: ParkAbbreviation) => {
  // re-use our query hooks whenever possible
  const { data, isLoading } = useStamps();
  return { data: data?.find((stamp) => stamp.code === code) || null, isLoading };
};
