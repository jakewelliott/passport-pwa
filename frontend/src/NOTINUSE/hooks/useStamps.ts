import { useQuery, useMutation, useQueryClient } from "@tanstack/react-query";
import type { Stamp } from "./types";
import { api } from "./api";
import { parkKeys } from "./useParks";

export const stampKeys = {
	all: ["stamps"] as const,
	lists: (parkId: number) => [...stampKeys.all, "list", parkId] as const,
	detail: (parkId: number, stampId: number) =>
		[...stampKeys.all, "detail", parkId, stampId] as const,
};

export const useStamps = (parkId: number) => {
	return useQuery<Stamp[], Error>({
		queryKey: stampKeys.lists(parkId),
		queryFn: () => api.getStamps(parkId),
		enabled: !!parkId,
	});
};

export const useCollectStamp = () => {
	const queryClient = useQueryClient();

	return useMutation<
		Stamp, // Success response type
		Error, // Error type
		{ parkId: number; stampId: number } // Variables type
	>({
		mutationFn: ({ parkId, stampId }) => api.collectStamp(parkId, stampId),
		onSuccess: ({ parkId }) => {
			// Invalidate related queries
			queryClient.invalidateQueries({ queryKey: stampKeys.lists(parkId) });
			queryClient.invalidateQueries({ queryKey: parkKeys.detail(parkId) });
		},
	});
};
