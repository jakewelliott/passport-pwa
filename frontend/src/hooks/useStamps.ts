import type { ParkAbbreviation } from '@/lib/mock/types';
import { useUser } from './queries/useUser';
import { api } from '@/lib/mock/api';
import { useQuery } from '@tanstack/react-query';
import { USER_LOADING_ID } from './queries/useUser';
/**
 * Gets all of the local user's stamps, empty array if user is not loaded
 * @returns The stamps for the user
 */
export const useStamps = () => {
  // re-use our query hooks whenever possible
  const { data: user } = useUser();
  // to respect rules of hooks we must always call this query hook, hence the USER_LOADING_ID
  return useQuery({
    queryKey: ['stamps', user?.id || USER_LOADING_ID],
    queryFn: () => api.getUserStampsByID(user?.id || USER_LOADING_ID),
  });
};

/**
 * Gets a single stamp for a user, null if stamp not collected yet
 * @param code The code of the park to get the stamp for
 * @returns The stamp for the user
 */
export const useStamp = (code: ParkAbbreviation) => {
  // re-use our query hooks whenever possible
  const { data, isLoading } = useStamps();
  return { data: data?.find((stamp) => stamp.code === code) || null, isLoading };
};
