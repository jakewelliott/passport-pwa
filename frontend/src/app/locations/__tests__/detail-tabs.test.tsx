import ParkInfoScreen from '@/app/locations/park-info';
import { usePark, useParks } from '@/hooks/queries/useParks';
import { useStamp, useStamps } from '@/hooks/queries/useStamps';
import { setupTestEnv } from '@/lib/testing/test-wrapper';
import { screen } from '@testing-library/react';
import { beforeAll, describe, expect, it } from 'vitest';

const routerProps = { initialEntries: ['/locations/CABE'] };
const { render, checkHook } = setupTestEnv({ routerProps });

describe('DetailTabs', () => {
    // ADAM: for some reason getting undefined error for stamps, skipping for now
    beforeAll(async () => {
        await checkHook(useParks, 'useParks');
        await checkHook(usePark, 'usePark', 'CABE');
        await checkHook(useStamps, 'useStamps');
        await checkHook(useStamp, 'useStamp', 45);
    });

    it('matches snapshot', () => {
        const { container } = render(<ParkInfoScreen />);
        expect(container).toMatchSnapshot();
    });

    it('fetches park activity for non-admin users', () => {
        render(<ParkInfoScreen />);
        // Verify components are rendered with park activity data
        expect(screen.getByTestId('location-contact')).toBeInTheDocument();
    });

    it('does not fetch park activity for admin users', () => {
        render(<ParkInfoScreen />);
        // Verify components are rendered without park activity data
        expect(screen.getByTestId('location-contact')).toBeInTheDocument();
    });
});
