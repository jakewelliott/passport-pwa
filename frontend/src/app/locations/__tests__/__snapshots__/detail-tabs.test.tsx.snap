// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DetailTabs > matches snapshot 1`] = `
<div>
  <div
    class="animate-pulse"
    data-testid="loading-placeholder"
  >
    <p
      class="text-center text-gray-500 text-sm"
    >
       Loading park...
    </p>
  </div>
</div>
`;
